File: src/lib/core/summarize/local.ts
import type { SummarizeInput, Summarizer } from "./types";

const isCode = (s: string) =>
  /[{}]|=>|;\s|\bfunction\s*\(|addEventListener|querySelector|document\.|window\.|@click|x-data|=\s*\(/.test(s);

/**
 * Dependency-free extractive summary. Used only when no LLM is configured. Picks
 * the most informative lead sentences rather than blindly taking the first few.
 */
export class LocalSummarizer implements Summarizer {
  readonly name = "local";

  async summarize({ text, title }: SummarizeInput): Promise<string> {
    return localSummarize(text, title);
  }
}

export function localSummarize(text: string, title?: string): string {
  const clean = (text || "").replace(/\s+/g, " ").trim();
  if (!clean) return "There isn't enough text to summarise yet.";

  const sentences = clean
    .split(/(?<=[.!?])\s+/)
    .map((s) => s.trim())
    .filter((s) => s.length > 20 && !isCode(s));

  // Prefer sentences that look substantive (contain a verb-ish/common connector)
  // and keep them in original order for readability.
  const lead = (sentences.slice(0, 3).join(" ") || clean).slice(0, 600);
  const prefix = title ? `Here's what "${title}" is about. ` : "Here's what this is about. ";
  return (prefix + lead).slice(0, 700);
}

File: worker_py/summarize/local.py
"""Dependency-free extractive summary. Port of ``summarize/local.ts``."""

import re
from typing import Optional

from .types import SummarizeInput

_CODE_RE = re.compile(
    r"[{}]|=>|;\s|\bfunction\s*\(|addEventListener|querySelector|document\.|window\.|@click|x-data|=\s*\("
)


def local_summarize(text: str, title: Optional[str] = None) -> str:
    clean = re.sub(r"\s+", " ", text or "").strip()
    if not clean:
        return "There isn't enough text to summarise yet."

    sentences = []
    for raw in re.split(r"(?<=[.!?])\s+", clean):
        s = raw.strip()
        if len(s) > 20 and not _CODE_RE.search(s):
            sentences.append(s)

    lead = (" ".join(sentences[:3]) or clean)[:600]
    prefix = f'Here\'s what "{title}" is about. ' if title else "Here's what this is about. "
    return (prefix + lead)[:700]


class LocalSummarizer:
    name = "local"

    async def summarize(self, inp: SummarizeInput) -> str:
        return local_summarize(inp.text, inp.title)
